lib/improver/spotdata/extract_data.py

```diff
@@ -0,0 +1,374 @@
+
+        """
+        latitudes = [site['latitude'] for site in sites.itervalues()]
+        longitudes = [site['longitude'] for site in sites.itervalues()]
+        utc_offsets = [site['utc_offset'] for site in sites.itervalues()]
+        site_ids = sites.keys()
+
+        indices, bd_ids, latitude, longitude, utc_offset = (
+            self._build_coordinates(
+                latitudes, longitudes, site_ids, utc_offsets))
+
+        # Add leading dimension for time.
+        data.resize(1, len(data))
+        result_cube = Cube(data,
+                           long_name=cube.name(),
+                           dim_coords_and_dims=[(cube.coord('time'), 0),
+                                                (indices, 1)],
+                           aux_coords_and_dims=[(latitude, 1),
+                                                (longitude, 1),
+                                                (utc_offset, 1),
+                                                (bd_ids, 1)],
+                           units=cube.units)
+
+        # Enables use of long_name above for any name, and then moves it
+        # to a standard name if possible.
+        result_cube.rename(cube.name())
+        return result_cube
+
+    def use_nearest(self, cube, sites, neighbours):
+        """
+        Simplest case, in which the diagnostic data value at the nearest grid
+        point, as determined by the chosen PointSelection method, is used for
+        the site.
+
+        Args:
+        -----
+        See process() above.
+
+        """
+        if (not cube.coord_dims(cube.coord(axis='y').name())[0] == 0 or
+                not cube.coord_dims(cube.coord(axis='x').name())[0] == 1):
+            raise Exception("Cube dimensions not as expected.")
+
+        data = cube.data[neighbours['i'], neighbours['j']]
+        return self.make_cube(cube, data, sites)
+
+    def orography_derived_temperature_lapse_rate(self, cube, sites, neighbours,
+                                                 orography, no_neighbours=9):
+        """
+        Crude lapse rate method that uses temperature variation and height
+        variation across local nodes to derive lapse rate. This is highly
+        prone to noise given the small number of points involved and the
+        variable degree to which elevation changes across the small number
+        of points.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            A cube of screen level temperatures at a single time.
+
+        sites/neighbours/no_neighbours : See process() above.
+
+        orography : numpy.array
+            Array of orography data on a grid that corresponds to the grid of
+            the the diagnostic cube.
+
+        Returns:
+        --------
+        Data extracted from the screen level temperature cube at spotdata sites
+        which has then been adjusted using a temperature lapse rate calculated
+        using local variations in temperature with orography.
+
```

Comment:
It would be great, if you could add in word form the relationship that you're establishing from the least squares fitting. Is there a reference that could be included for this type of lapse rate calculation? What about [Stahl et al., 2006](https://www.researchgate.net/publication/228933348_Comparison_of_Approaches_for_Spatial_Interpolation_of_Daily_Air_Temperature_in_a_Large_Region_With_Complex_Topography_and_Highly_Variable_Station_Density)?